src/Sparrow/Json/JsonContextPoolBase.cs

```diff
@@ -388,32 +152,90 @@ public void Dispose()
+            TimeSpan idleTime = TimeSpan.FromMinutes(5);
+
+            if (_globalQueue.IsEmpty)
             {
-                context.Dispose();
+                // we have nothing in global, let's check if we can clear the per core
+                foreach (var current in _perCoreContexts)
+                {
+                    for (var index = 0; index < current.Length; index++)
+                    {
+                        var context = current[index];
+                        if(context == null)
+                            continue;
+
+                        if (currentTime - context.InPoolSince > idleTime)
+                            continue;
+
+                        if (!context.InUse.Raise())
+                            continue;
+
+                        Interlocked.CompareExchange(ref current[index], null, context);
+                        context.Dispose();
+                    }
+                }
+
                 return;
             }
-            while (true)
+
+            while (_globalQueue.TryPeek(out var context))
             {
-                var current = threadHeader.Head;
-                if (current == ContextStack.HeaderDisposed)
+                if (currentTime - context.InPoolSince < idleTime)
                 {
-                    context.Dispose();
-                    return;
+                    break;
                 }
-                var newHead = new StackNode<T> { Value = context, Next = current };
-                if (Interlocked.CompareExchange(ref threadHeader.Head, newHead, current) == current)
+
+                if (_globalQueue.TryDequeue(out context) == false)
+                    break;
+
+                if (currentTime - context.InPoolSince > idleTime)
+                {
+                    if (context.InUse.Raise())
+                        context.Dispose();
+                }
+                else
+                {
+                    // was checked out, meaning there is activity, we are done now
+                    _globalQueue.Enqueue(context);
+                    break;
+                }
+            }
+        }
+
+        private void Push(T context)
+        {
+            int currentProcessorId = CurrentProcessorIdHelper.GetCurrentProcessorId() % _perCoreContexts.Length;
+            var core = _perCoreContexts[currentProcessorId];
+
+            for (int i = 0; i < core.Length; i++)
+            {
+                if (core[i] != null)
+                    continue;
+                if (Interlocked.CompareExchange(ref core[i], context, null) == null)
                     return;
             }
+
+            if (core.Length < 64)
+            {
+                var newCore = new T[core.Length * 2];
+                Array.Copy(core, 0, newCore, 0, core.Length);
+                newCore[core.Length] = context;
+                if (Interlocked.CompareExchange(ref _perCoreContexts[currentProcessorId], newCore, core) == core)
```

Comment:
I think you might leak contexts here. 
If someone put the context back, after you copied it, but before the exchange.

Reply:
Good point, I think this is going to be really rare thing, and will be covered by the finalizer. But will think on how to resolve this.